### DISCUSSION TITLE
Setting compiler options

### DISCUSSION THREAD

**Discussion by TheComamba** [texstudio-org/texstudio#3058] [Q&A] [answered]
Hi everyone!
Question:
In the gui, is it possible to specify command line options to be passed to the tex compiler?

Background:
I want to use lualatex to load a shared library. It knows the ffi (foreign function interface) module, but it can (at least on my ubuntu machine) only access the ffi.load function if I compile using ``lualatex --enable-write18 <file.tex>``, I guess because it is secure by default. It would be nice to keep compiling using this wonderful gui.

Thanks a lot in advance!

**Comment by muzimuzhi** [comment, marked as answer]
I used to execute `-shell-escape` variants by adding a user command (Configure TeXstudio - "Build" tab - User commands) then add a new button linked to the automatically created menu "Tools - User - <name of just created user command>".  The disadvantage is that the button is hard-coded to an engine, for example `pdflatex`.  Thus if I want to switch to another engine it seems I have to add another button.

Now I use magic comment (menu "LaTeX - Add magic comments..."; also called ["advanced header"](https://texstudio-org.github.io/advanced.html#advanced-header-usage) in the new manual) to specify the engine and options.
```
% !TeX program = lualatex -synctex=1 -interaction=nonstopmode -shell-escape %.tex | txs:///view
```
On macOS you may need https://github.com/texstudio-org/texstudio/issues/143#issuecomment-1498207044.  The disadvantage is that I have to repeat the default options `-synctex=1 -interaction=nonstopmode`.

TeXstudio supports [subcommands argument modifiers](https://texstudio-org.github.io/configuration.html#advanced-configuration-of-the-build-system), for example `txs:///foobar --xyz` will add option `--xyz` and `txs:///foobar/{--xyz}` will remove option `--xyz`.  But it seems they're not very useful, see #3062.

**Comment by TheComamba** [comment]
Together with your hack from [#3062](https://github.com/texstudio-org/texstudio/discussions/3062) this is exactly what I had hoped for.
Thank you so much!

P.S.: The magic comment was not recognised on my machine - but that's a problem for another day.

**Comment by muzimuzhi** [comment]
@TheComamba As corrected in https://github.com/texstudio-org/texstudio/pull/3061#issuecomment-1500698027 by @octaeder, the right syntax should be
```
% !TeX TXS-program:compile = txs:///lualatex/{%.tex} -shell-escape "%.tex"
```
`% !TeX program = ...` only supports a value in one of “latex”, “pdflatex”, “xelatex”, “lualatex” and “user<i>n</i>”, see manual https://texstudio-org.github.io/advanced.html#advanced-header-usage.
